### DISCUSSION TITLE
HyperSPI on ESP32 MH-ET Live (CH9102x)

### DISCUSSION THREAD

**Discussion by ghost** [awawa-dev/HyperSPI#12] [Show and tell]
Good day dear @awawa-dev

I ordered all the necessary components for the assembly of the highlight.
Bought on your recommendation:
- _Raspberry pi 4 1Gb_
- _Ezcoo ‎EZ-SP12HAS_
- _USB grabber MS2130_
- _ESP32 MH-ET Live (MiniKIT v2.0)_
- _Level Shifter_ (SN74AHCT125N)
- _WS2812B_

Dupont connecting wires will not exceed **10 cm** (if I'm not mistaken should not exceed 15)
All modules will be located in one box (as in [your blog](https://www.hyperhdr.eu/2023/02/ultimate-guide-on-how-to-build-led.html)).

**Comment by awawa-dev** [comment]
Hey
1. The level shifter from the diagram is not intended for Neopixel LEDs. You should use this: https://github.com/awawa-dev/HyperHDR/wiki/Level-Shifter
2. Precompiled firmware is already available in the release section, you don't need to compile your own: https://github.com/awawa-dev/HyperSPI/releases Unless you want to change the pinout for some reason, but that's not recommended. If you choose to do this, your approach is correct, with these lines to be commented out only in the HyperSPI_esp32_Neopixel.ino schema.
3. You need a common ground between the esp32 (or levelshifter) and the LED power supply, which is included in your schematic. But you also need common ground between ESP and Rpi. I recommend using a short ground between their pinouts like you connect their SPI interfaces. The USB cable between the Rpi and the ESP should also provide this, but in some cases it may not be enough for a stable SPI connection.

**Reply by ghost** [reply]
Thanks for the quick response!
> The level shifter from the diagram is not intended for Neopixel LEDs

It turns out that I ordered this ready-made converter in vain))
I saw it [in this video](https://www.youtube.com/watch?v=Q9mLitVxF3c) with the ws2812b tape, I thought that you can use it here too.

> I recommend using a short ground between their pinouts like you connect their SPI interface

Understood. I will connect with an additional short wire.

> Precompiled firmware is already available in the release section

I just didn't know what DATA output was in the finished firmware and whether the first LED was disabled in it.

**Reply by awawa-dev** [reply]
> I just didn't know what DATA output was in the finished firmware and whether the first LED was disabled in it.

They come with with all LEDs enabled by default. The feature to disable the first LED is not needed anymore because we can do it in HyperHDR: https://github.com/awawa-dev/HyperHDR/pull/379 The data out in on GPIO2

>It turns out that I ordered this ready-made converter in vain))
I saw it [in this video](https://www.youtube.com/watch?v=Q9mLitVxF3c) with the ws2812b tape, I thought that you can use it here too.

It's just too slow for Neopixel, in some cases it can only make things worse than not using any level shifter. You can try it but I don't gurantee anything, SN74AHCT125N  is proven solution used in all my projects and it's also recommended by WLED. And you can also find there such warning:
![obraz](https://user-images.githubusercontent.com/69086569/215278492-f3733a31-24dd-4cb6-b331-1f63fa110ad5.png)

**Reply by ghost** [reply]
> SN74AHCT125N is proven solution used in all my projects and it's also recommended by WLED

That's why I asked my question to hear your expert opinion (since I don't know much about it myself).
Anything that is stable, I will use))

**Reply by ghost** [reply]
Updated my layout for assembly, made all the changes with which you helped me.
I do not know if a capacitor is needed for the LEDs on the power line after the power supply.
Maybe he's redundant.)

Now it remains to wait for the components and start the assembly.

**Reply by awawa-dev** [reply]
Great plan! I'll pin it later. Minor notes:
- the capacitor is rather not necessary
- if you connect ESP with Rpi using USB cable (including data lines, not only power supply) your setup will become compatible also with HyperSerialESP32: why should we limit ourselves? ;) CH9102x on this board supports even 4Mb speed.

Probably in the near future I will update HyperSPI to catch up with HyperSerialESP32 features (platformio & I2S prefill support etc) so you can check back here soon.

**Reply by ghost** [reply]
> CH9102x on this board supports even 4Mb speed

Therefore, I purchased a board with just such a chip (so that it is suitable for all types of connection, universal).
I saw that you wrote that there are many fakes with the CP2104 chip, and did not try my luck)

**Reply by awawa-dev** [reply]
Usually ch340 on ESP8266 is counterfeit and reported cases are increasing. ESP32 boards are not affected yet, but there are some poorly designed models that may not work properly even if it's CH9102x  [https://github.com/awawa-dev/HyperHDR/discussions/401#discussioncomment-4653269](https://github.com/awawa-dev/HyperHDR/discussions/401#discussioncomment-4653269) But MH ET Live is solid and proven solution.

**Reply by ghost** [reply]
I was just looking at this comment when I made an order and immediately refused such boards).
I also saw in one video from one of the buyers about such ESP32, many manufacturers save on assembly and do not install a protective diode (the installation location is indicated by a red arrow in the picture). Instead, solder zero resistance. As a result, this gives us the fact that if the power supply of the board to the 'VIN' pin is more than 5v (voltage surges, for example at some point), then the device connected via micro-Usb will burn out as this unstabilized voltage will fall on it.
<details>
  <summary>Defective boards</summary>
  
  ![ESP32](https://user-images.githubusercontent.com/85223482/215340288-ee5196e5-5dc6-4a84-a050-145cd382ff64.png)
  
</details>

**Reply by awawa-dev** [reply]
It seems so: in review section with "1 star" there is a high-resolution photo and a zero resistor in this place. By the way HyperSPI is almost ready to migrate to platformio and has the same features as HyperSerialESP32 but with lower latency, using lighter SPI instead of USB serial communication, output frame rate is more stable. I think I need another week or two before merging, because there are a lot of changes in my fork of Neopixelbus library: [PR: work in progress](https://github.com/awawa-dev/HyperSPI/pull/13) It works also ESP32-S2 mini very well. Since they share the same wemos shield pinout format, a prototype board could be design in the way to allow replace/change all these ESP8266/ESP32/ESP32-S2 without any hardware modification on the board (only pinout in the firmware)

**Reply by awawa-dev** [reply]
Also tested new version on ESP8266/ESP32/ESP32-S2: MISO connection is not needed. I updated the info in the PR. But short common ground is very important: it didnt work well if ESP and Rpi shared it over USB cable.

**Reply by ghost** [reply]
> Probably in the near future I will update HyperSPI to catch up with HyperSerialESP32 features (platformio & I2S prefill support etc) so you can check back here soon.

With the introduction of support for parallel multi-segment mode, the circuit can now be slightly changed.
On long sections of the tape, it is now possible to arrange not only a power line on both sides, but also a data line.

<details>
  <summary>Multi-segment mode</summary>

  ![HyperSPI](https://user-images.githubusercontent.com/85223482/222923979-f344349a-1f8b-4195-94ca-51721923359e.png)

</details>

This way we can avoid the voltage drop at the end of the line and additionally double the update rate.

**Reply by awawa-dev** [reply]
Just pay attention to the right segment: the arrows on the LED strip should be directed in the opposite direction, and consequently: +5V and ground should also change their position accordingly on the second segment. Such a configuration as in the current diagram is of course also possible, but the input data for the second segment should be in the middle of the upper edge.

**Reply by ghost** [reply]
It's just that in the Fritzing editor it turned out to draw like that :)
From the commentary to the release, I understood how the ribbon should be located.
It was just easier to redraw this part in the editor in this way.
Updated the picture.

**Reply by awawa-dev** [reply]
Excellent! May I ask you as the author to create two PR and add this image to readme.md in HyperSPI (around 154 or 169 line) and HyperSerialESP32 (around 91 or 107 line). Add new lines around the pictures if necessary to separate it from the text.

**Reply by ghost** [reply]
Ready. For a long time, of course, I did not use Git, I again remembered how to correctly send a signed commit :)

**Reply by awawa-dev** [reply]
No problem :) PR are already merged.

**Comment by ghost** [comment]
Good afternoon dear @awawa-dev

I assembled a level shifter and an Esp32 on the same board like you have in your blog.
In some [manuals](https://electronics.stackexchange.com/questions/102050/what-is-the-purpose-of-adding-a-300-ohm-to-500-ohm-resistor-on-the-ws2812b-neopi) for LED strips, it is recommended to install resistance on the Data line.

<details>
  <summary> </summary>
  
Adding a 300 to 500 Ohm resistor between your microcontroller's data pin and the data input on the first NeoPixel can help prevent voltage spikes that might otherwise damage your first pixel. Please add one between your micro and NeoPixels!
  
</details>

Is it really necessary or everything works without it.
I already have a 470 Ohm resistance, I don’t know if I need to install it or not)
